File: problems/1145_binary_tree_coloring_game/Solution.java
class Solution {
    int leftCount = 0, rightCount = 0;
    public boolean btreeGameWinningMove(TreeNode root, int n, int x) {
        find(root, x);
        int parentCount = n - leftCount - rightCount - 1;
        return Math.max(Math.max(leftCount, rightCount), parentCount) > n / 2;
    }
    private int count(TreeNode node) {
        if (node == null) return 0;
        return 1 + count(node.left) + count(node.right);
    }
    private void find(TreeNode node, int x) {
        if (node == null) return;
        if (node.val == x) {
            leftCount = count(node.left);
            rightCount = count(node.right);
            return;
        }
        find(node.left, x);
        find(node.right, x);
    }
}

File: problems/1145_binary_tree_coloring_game/solution.py
from shared.python.data_structures import TreeNode


class Solution:
    def btreeGameWinningMove(self, root: TreeNode, n: int, x: int) -> bool:
        left_count = right_count = 0
        def count(node: TreeNode) -> int:
            if not node:
                return 0
            return 1 + count(node.left) + count(node.right)
        def find(node: TreeNode) -> None:
            nonlocal left_count, right_count
            if not node:
                return
            if node.val == x:
                left_count = count(node.left)
                right_count = count(node.right)
                return
            find(node.left)
            find(node.right)
        find(root)
        parent_count = n - left_count - right_count - 1
        return max(left_count, right_count, parent_count) > n // 2
